File: src/0347-top-k-frequent-elements/solution.py
from collections import defaultdict
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashmap = defaultdict(int)
        for i, ele in enumerate(nums):
            hashmap[ele] += 1
        # since we are sorting based on occurence, we can use buckets based on length of nums
        # if we were sorting based on min num and max num, then its a bit different
        buckets = [[] for _ in range(len(nums) + 1)]
        for num, occurence in hashmap.items():
            buckets[occurence].append(num)

        # Now we can iterate backwards, with highest occurence at the end of the array
        res = list()
        for i in range(len(buckets) - 1, -1, -1):
            if k == 0:
                return res
            if buckets[i]:
                res += buckets[i]
                # there could be more than one element in the bucket
                k -= len(buckets[i])

        return res

File: src/0347-top-k-frequent-elements/solution.ts
function topKFrequent(nums: number[], k: number): number[] {
    const occurence = new Map<number, number>();
    for (const n of nums) {
        if (occurence.has(n)) {
            occurence.set(n, occurence.get(n) + 1);
        } else {
            occurence.set(n, 1);
        }
    }

    // Our bucket sort is like so
    // number of occurences array => contains number for that occurence
    // create as many buckets as there is elements
    // Step 2: Create the buckets array (each bucket should be a separate empty array)
    const buckets = Array.from({ length: nums.length + 1 }, () => [])

    for (const [number, occured] of occurence) {
        buckets[occured].push(number)
    }

    // Step 3, iterate backwards, as last few buckets are the top occuring elements
    const res = [];
    console.log(buckets)
    for (let i = buckets.length - 1; i >= 0; i--) {
        if (res.length === k) {
            return res;
        }
        if (buckets[i].length > 0) {
            res.push(...buckets[i])
        }
    }
    return res;
};
